core/src/main/java/org/infinispan/remoting/inboundhandler/NonTotalOrderTxPerCacheInboundInvocationHandler.java

```diff
@@ -42,4 +123,288 @@ protected boolean isTraceEnabled() {
       return trace;
    }
 
+   protected final BlockingRunnable createReadyActionAwareRunnable(CacheRpcCommand command, Reply reply, int commandTopologyId,
+                                                                   boolean waitTransactionalData, boolean onExecutorService,
+                                                                   ReadyAction readyAction) {
+      final TopologyMode topologyMode = TopologyMode.create(onExecutorService, waitTransactionalData);
+      if (onExecutorService && readyAction != null) {
+         readyAction.init();
+         readyAction.addListener(remoteCommandsExecutor::checkForReadyTasks);
+         return new DefaultTopologyRunnable(this, command, reply, topologyMode, commandTopologyId) {
+            @Override
+            public boolean isReady() {
+               return super.isReady() && readyAction.isReady();
+            }
+         };
+      } else {
+         return new DefaultTopologyRunnable(this, command, reply, topologyMode, commandTopologyId);
+      }
+   }
+
+   private ReadyAction createReadyActionForLockControlCommand(LockControlCommand command) {
+      if (command.hasSkipLocking()) {
+         return null;
+      }
+      Collection<Object> keys = command.getKeysToLock();
+      if (keys.isEmpty()) {
+         return null;
+      }
+      final List<Object> keysToLock = new ArrayList<>(keys.size());
+      filterByPrimaryOwner(keys, keysToLock);
+      if (keysToLock.isEmpty()) {
+         return null;
+      }
+      final long timeoutMillis = command.hasZeroLockAcquisition() ? 0 : configuration.locking().lockAcquisitionTimeout();
+      final RemoteTxInvocationContext context = command.createContext();
+      keysToLock.forEach(context::addLockedKey);
+      if (keysToLock.size() == 1) {
+         return new SingleKeyReadyAction(context, timeoutMillis, keysToLock.get(0));
+      }
+      return new MultipleKeysReadyAction(context, timeoutMillis, keysToLock);
+   }
+
+   private ReadyAction createReadyActionForPrepareCommand(PrepareCommand command) {
+      if (command.hasSkipLocking()) {
+         return null;
+      }
+      Collection<Object> keys = command.getKeysToLock();
+      if (keys.isEmpty()) {
+         return null;
+      }
+      final List<Object> keysToLock = new ArrayList<>(keys.size());
+      filterByPrimaryOwner(keys, keysToLock);
+      if (keysToLock.isEmpty()) {
+         return null;
+      }
+      final long timeoutMillis = command.hasZeroLockAcquisition() ? 0 : configuration.locking().lockAcquisitionTimeout();
+      final RemoteTxInvocationContext context = command.createContext();
+      if (context == null) {
+         return null;
+      }
+      keysToLock.forEach(context::addLockedKey);
+      if (keysToLock.size() == 1) {
+         return new SingleKeyReadyAction(context, timeoutMillis, keysToLock.get(0));
+      }
+      return new MultipleKeysReadyAction(context, timeoutMillis, keysToLock);
```

Comment:
probably with all the refactor I have missed that. I'll double check and if it is the case I'll remove one them :)